README.md

# Computer vision with webcam

- Zhijin Li

This project contains implementations of various computer vision
applications for webcam. Models used in this repo run in inference
mode, with weights obtained from official online public
repositories. The possibility of retraining or finetuning the models
using customized dataset is out-of-scope of this repo.

All codes were developped and tested on a 15-inch MacBook Pro
(mid-2018). An effort was made so that the applications in this repo
run at at least around 30 frames per second (fps). Content of this
repo is detailed Hereunder.


----------


## Repo structure

- `cam-*.py` files refer to different applications. It can be run in a
  terminal by typing `./cam-*.py`.
- `lib` folder contains source code for the applications.
- `data` folder contains images and files describing models and
  datasets used in this repo. Some large data, such as model weights
  and additional images will be downloaded into this folder when
  certain applications are launched for the first time.
- `ChangeLog`: change tracker.
- `AUTHORS`
- `LICENSE`
- `README.md`


----------


## Camera capture utilities

A utility for real-time camera capturing is provided in `lib/utils`
under the name `capture.py`. It can be launched in a terminal. In this
case a demo frame capturing real-time camera feed will pop up. FPS
measurement is printed at the top-left corner of the frame. This file
is also imported as a module in other applications, providing utility
APIs related to threaded frame capturing, frame trimming and
application-specific frame modifications.

### Dependencies

- `opencv-python`
- `numpy`


----------


## Object classification with `cam-classify`

Classifiy object from real-time webcam feed.

<img
    src = "./data/img/cam-classify-sample1.png"
    alt-text = "cam-classify example 1"
    width = "400"
    height = "440"
    />
<img
    src = "./data/img/cam-classify-sample2.png"
    alt-text = "cam-classify example 2"
    width = "400"
    height = "440"
    />

This application uses
[`MobileNet`](https://arxiv.org/pdf/1704.04861.pdf) from the official
[`tf.keras.applications`](https://keras.io/applications/#mobilenet)
with weights pre-trained on `ImageNet`.

### Usage

Run

    ./cam-classify.py

in a terminal.

### Dependencies

- `opencv-python`
- `numpy`
- `tensorflow`


----------


## Object detection with `cam-detect`

Detect objects from real-time webcam feed.

<img
    src = "./data/img/cam-detect-sample.png"
    alt-text = "cam-classify example 1"
    width = "400"
    height = "440"
    />

This application implements
[`YOLOv3-tiny`](https://pjreddie.com/media/files/papers/YOLOv3.pdf)
architecture in `pytorch` for object detection. The model weights was downloaded
from the official [`darknet`
website](https://pjreddie.com/darknet/yolo/). An effort was made to
replicate the result of the original `darknet` implementation,
with maximum absolute difference between detection feature-maps from
two implementation bounded by $10^{-4}$, when the darknet
pre-processed images are used as input.

### Usage

Run

    ./cam-detect.py <--run_test> <--use_darknet_boxes>

in a terminal.

- `--run_test` is an optional argument. When specified, the
  application will run on a series of `darknet` test images and
  display the results.
- `--use_darknet_boxes` is an optional argument and can be added in
  additiona to `--run_test`. When specified, the application will use
  `darknet` pre-processed (letter box transformed) images as input. An
  utility program `make_letter_boxes` is provided under `tool/darknet`
  to transform color images to Darknet style letter boxes. To use this
  utility, go to `tool/darknet`, run

        make

  to compile the program. Then run

        ./make_letter_boxes ../../data/darknet-test-images/ jpg

  The program will save letter box transformed images under the
  directory `data/darknet-test-images`.

### Dependencies

- `opencv-python`
- `numpy`
- `torch`, `torchvision`
- `requests`
- [`darknet`](https://pjreddie.com/darknet/install/) compiled &
  installed (only for `make_letter_boxes` utility)


----------


### Face-switching with `cam-switch`


----------


### Pose detection with `cam-pose`


ChangeLog
2019-01-06  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* README.md: Add entries for on-going implementations.

2018-12-24  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* README.md: Refine documentation.

	* tool/darknet/make_letter_boxes.c: Provide utility program for
	darknet letter box transform.

2018-12-22  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* README.md: Update documentation.

2018-12-17  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* README.md: Update documentation. Add documentation images.
	Change example image size.

	* lib/utils/capture.py: Fix video cam demo.

2018-12-16  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* cam-detect.py: Add test option with darknet letter boxes.

	* lib/utils/utils.py: Refactorize url download.

	* cam-classify.py: Change MobileNet input size.

2018-12-13  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/utils/utils.py (download_yoolov3tiny_weights): Add weight
	download API.

	* cam-detect.py: Implement real-time cam detection.

	* lib/yolov3tiny.py (detect_color_img): Refactorize detection
	APIs.

2018-12-12  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/utils/capture.py: Improve frame trimming.

	* cam-detect.py: Test on Darknet images.

	* lib/utils/utils.py: Implement batch image loader.

	* lib/utils/capture.py: Correct detection drawing bounding error.

2018-12-11  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/utils/utils.py (nms): Implement nms.

2018-12-10  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/utils/capture.py: Adapt detection frame to letterbox and
	corrected images.
	Bug fix in detection frame.

	* lib/utils/utils.py: Implement letterbox transformation and bbox
	correction.

	* lib/utils/capture.py: Make better prediction frame.

2018-12-09  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/utils/utils.py: Create detection frame.

	* lib/yolov3tiny.py: Refactorize conv block.

2018-12-08  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/utils/utils.py: Add BR channel swap and channel_first
	reordering in make_predict_inp.

	* lib/yolov3tiny.py: Fix same padding value for max pooling
	layer.
	Refactorize set_weights interface.

2018-12-06  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* cam-classify.py: Fix cam-classify input image channel ordering.

2018-12-04  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/yolov3tiny.py (YOLODetect.__transform_bbox_centers): Bugfix
	set model to eval mode.

2018-12-03  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/yolov3tiny.py (YOLO.forward): Fix batch norm epsilon.

2018-12-02  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/yolov3tiny.py: Implement Darknet weight parser.

2018-12-01  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/yolov3tiny.py: Fix YOLO conv layer bias.

2018-11-29  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/yolov3tiny.py: Fix detection return.

2018-11-28  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/yolov3tiny.py (PaddingSame): Fix max pooling padding.
	(NearestInterp): Fix nearest neighbor upsample layer.

2018-11-27  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/yolov3tiny.py: Fix conv layer padding.

2018-11-26  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/yolov3tiny.py: Implement detection layer transform.

2018-11-25  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/yolov3tiny.py: Fix routing layers.
	Implement detection layer.

2018-11-21  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/yolov3tiny.py: Fix YOLO constructor.

2018-11-19  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* LICENSE (met): Update license.

	* lib/yolov3tiny.py: Implement YOLO model main loop.

2018-11-18  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/yolov3tiny.py: Implement oute layer.

2018-11-13  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/yolov3tiny.py (YOLO.__make_upsample): Modify repo
	structure. Add YOLO config parser.

2018-11-12  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* lib/utils/utils.py: Restructure repo. Add YOLO detection project
	entries.

2018-10-30  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* utils/utils.py: Fix image path param in load_img.

2018-10-29  Zhijin Li  <jonathan.zj.lee@gmail.com>

	* README.md: Initialize repo.


utils/mobilenet.py
#! /usr/bin/env python3
# Copyright (C) 2018  Zhijin Li

# All rights reserved.

# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:

#     * Redistributions of source code must retain the above copyright
# notice, this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above
# copyright notice, this list of conditions and the following disclaimer
# in the documentation and/or other materials provided with the
# distribution.

# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

## ---------------------------------------------------------------------------
##
## File: mobilenet.py for Cam-Vision
##
## Created by Zhijin Li
## E-mail:   <jonathan.zj.lee@gmail.com>
##
## Started on  Sun Oct 28 21:06:23 2018 Zhijin Li
## Last update Mon Oct 29 22:34:54 2018 Zhijin Li
## ---------------------------------------------------------------------------


import numpy as np
import tensorflow as tf


def __load_mobilenet_base(width_mult):
  """

  Load MobileNet with weights pretrained
  ImageNet.

  width_mult: float
  Width multiplier used for decreasing the
  number of filters in conv layers.

  Returns
  ----------
  tf.keras.models.Model
  Loaded MobileNet model.

  """
  res = tf.keras.applications.mobilenet.MobileNet(
    include_top=True,
    weights='imagenet',
    alpha=width_mult,
    pooling='avg')
  return res


def __load_mobilenet_empty(width_mult, pooling):
  """

  Load MobileNet without weights.

  Parameters
  ----------
  width_mult: float
  Width multiplier used for decreasing the
  number of filters in conv layers.

  pooling: str
  The final global pooling type. Either
  `global_avg` or `global_max`.

  Returns
  ----------
  tf.keras.models.Model
  Loaded MobileNet model.

  """
  if pooling not in ['global_avg', 'global_max']:
    raise AttributeError(
      'pooling must be global_avg or global_max.')

  res = tf.keras.applications.mobilenet.MobileNet(
    input_shape=(None,None,3),
    include_top=False,
    weights=None,
    alpha=width_mult,
    depth_multiplier=1, # Current keras impl seems to only works for 1.
    pooling=pooling.split('_')[-1])
  return res


def __get_global2dpooling(model):
  """

  Get the first global 2D pooling layer.

  Parameters
  ----------
  model: tf.keras.models.Model
  The input keras model.

  Returns
  ----------
  The first global pooling layer.

  """
  __lay = None
  for __l in model.layers:
    if (isinstance(__l,tf.keras.layers.GlobalMaxPooling2D) or
        isinstance(__l,tf.keras.layers.GlobalAveragePooling2D)):
      __lay = __l
  return __lay


def __customize_mobilenet(base_model, empty_model):
  """

  Customize MobileNet for inference mode with
  arbitrary input image size.

  Parameters
  ----------
  base_model: tf.keras.models.Model
  The input base model with weights pretrained
  on ImageNet.

  empty_model: tf.keras.models.Model
  The input empty model without pretrained weights,
  configured for arbitrary input image size.

  Returns
  ----------
  tf.keras.models.Model
  The customized model, with weights copied
  from the base model.

  """
  intercept = empty_model.output
  __s = empty_model.layers[-1].get_output_shape_at(0)[-1]
  intercept = tf.keras.layers.Reshape((1,1,__s))(intercept)
  intercept = tf.keras.layers.Conv2D(
    name='prediction', filters=1000, kernel_size=(1,1),
    strides=(1,1), padding='same',activation='softmax')(
      intercept)
  intercept = tf.keras.layers.Flatten()(intercept)
  model = tf.keras.models.Model(
    inputs=empty_model.input, outputs=intercept)

  for idx in range(len(empty_model.layers)):
    model.layers[idx].set_weights(
      base_model.layers[idx].get_weights())
  __w = base_model.get_layer('conv_preds').get_weights()
  model.get_layer('prediction').set_weights(__w)
  return model


def load_mobilenet_anysize(width_mult, pooling):
  """

  Load a MobileNet for images of any size.

  Copy weights of a loaded base MobileNet
  to an anysize empty MobileNet and return it.

  Parameters
  ----------
  width_mult: float
  Width multiplier used for decreasing the
  number of filters in conv layers.

  pooling: str
  The final global pooling type. Either
  `global_avg` or `global_max`.

  Returns
  ----------
  tf.keras.models.Model
  Loaded MobileNet model.

  """
  print(
    ('\nloading MobileNet with\n'
     '- {:25s}: {}\n- {:25s}: {}\n').format(
       'width multiplier', width_mult,
       'pooling type', pooling))
  __base = __load_mobilenet_base(width_mult)
  __res = __load_mobilenet_empty(width_mult,pooling)
  __res = __customize_mobilenet(__base,__res)
  print('model loaded.')
  return __res
